python/ray/data/_internal/planner/exchange/shuffle_task_spec.py

```diff
@@ -33,22 +42,37 @@ def map(
         idx: int,
         block: Block,
         output_num_blocks: int,
+        target_max_block_size: int,
         upstream_map_fn: Optional[Callable[[Iterable[Block]], Iterable[Block]]],
         random_shuffle: bool,
         random_seed: Optional[int],
     ) -> List[Union[BlockMetadata, Block]]:
-        # TODO: Support fusion with other upstream operators.
         stats = BlockExecStats.builder()
         if upstream_map_fn:
-            mapped_blocks = list(upstream_map_fn([block]))
-            if len(mapped_blocks) > 1:
-                builder = BlockAccessor.for_block(mapped_blocks[0]).builder()
-                for b in mapped_blocks:
-                    builder.add_block(b)
-                block = builder.build()
-            else:
-                block = mapped_blocks[0]
+            # TODO: Support dynamic block splitting in
+            # all-to-all ops, to avoid having to re-fuse
+            # upstream blocks together.
+            upstream_map_iter = upstream_map_fn([block])
+            mapped_block = next(upstream_map_iter)
+            builder = BlockAccessor.for_block(mapped_block).builder()
+            builder.add_block(mapped_block)
+            for mapped_block in upstream_map_iter:
+                builder.add_block(mapped_block)
+            # Drop the upstream inputs to reduce memory usage.
+            del mapped_block
+            block = builder.build()
         block = BlockAccessor.for_block(block)
+        if block.size_bytes() > target_max_block_size:
+            logger.get_logger().warn(
+                "Input block to map task has size "
+                f"{block.size_bytes() // (1024 * 1024)}MiB, which exceeds "
+                "DataContext.get_current().target_shuffle_max_block_size="
+                f"{target_max_block_size // (1024 * 1024)}MiB. "
+                "This can lead to out-of-memory errors and can happen "
+                "when map tasks are fused to the shuffle operation. "
+                "To prevent fusion, call Dataset.materialize() on the "
+                "dataset before shuffling."
```

Comment:
Setting different resources can also prevent fusion, without having to materialize all data. 
Another workaround is to increase the parallelism to make each map op finer-grained.

Reply:
I think materializing is fine here because you have to materialize all data for an all-to-all op anyway.

Reply:
makes sense.